File: packages/sdk/plugin/contributions/manager.ts
/**
 * Contribution Manager - Central facade for all plugin contribution registries.
 *
 * Manages specialized per-type registries and handles cross-cutting concerns
 * such as override policy conflict resolution. Implements ContributionRegistrar
 * directly, eliminating the need for separate adapter classes.
 */

import type { BaseFormatter } from "../../services/llm/formatters/base.js";
import type { NodeHandlerFn } from "../../workflow/execution/handlers/node-handlers/index.js";
import type { ContributionType } from "./contribution-type.js";
import type { ExecutionMiddleware, MiddlewarePhase } from "./middleware.types.js";
import type {
  PluginNodeHandler,
  PluginToolExecutor,
  PluginLLMFormatter,
  PluginEventHandler,
  PluginHookHandler,
  PluginExecutionMiddleware,
  PluginLLMRequest,
  PluginLLMResponse,
} from "./abstractions.js";
import type {
  NodeTypeRegistrar,
  ToolTypeRegistrar,
  LLMProviderRegistrar,
  FormatterRegistrar,
  EventHandlerRegistrar,
  HookHandlerRegistrar,
  MiddlewareRegistrar,
  ContributionRegistrar,
} from "./registration.js";
import { OverridePolicy } from "../config.js";
import { createContextualLogger } from "../../utils/contextual-logger.js";
import type { IToolExecutor } from "../../services/tools/core/interfaces.js";
import type { Tool, ToolExecutionOptions, ToolExecutionResult } from "@wf-agent/types";

import { NodeTypeRegistry } from "./registries/node-type-registry.js";
import { ToolTypeRegistry, type IToolExecutorConstructor } from "./registries/tool-type-registry.js";
import { LLMProviderRegistry } from "./registries/llm-provider-registry.js";
import { FormatterRegistry } from "./registries/formatter-registry.js";
import { EventHandlerRegistry, type EventHandler } from "./registries/event-handler-registry.js";
import { HookHandlerRegistry, type HookHandler } from "./registries/hook-handler-registry.js";
import { MiddlewareRegistry } from "./registries/middleware-registry.js";

// ============================================================
// Type Conversion Helpers
// ============================================================

/**
 * Convert a PluginLLMFormatter to a BaseFormatter.
 */
function toBaseFormatter(formatter: PluginLLMFormatter): BaseFormatter {
  return {
    format: async (request: PluginLLMRequest): Promise<PluginLLMResponse> => {
      return formatter.format(request);
    },
  } as unknown as BaseFormatter;
}

// ============================================================
// Contribution Manager
// ============================================================

/**
 * ContributionManager - Central manager for all plugin contributions.
 *
 * Implements ContributionRegistrar directly, so plugins interact with
 * this single instance through the registrar interface.
 */
export class ContributionManager implements ContributionRegistrar {
  private logger = createContextualLogger({ component: 'ContributionManager' });

  private nodeTypeRegistry = new NodeTypeRegistry();
  private toolTypeRegistry = new ToolTypeRegistry();
  private llmProviderRegistry = new LLMProviderRegistry();
  private formatterRegistry = new FormatterRegistry();
  private eventHandlerRegistry = new EventHandlerRegistry();
  private hookHandlerRegistry = new HookHandlerRegistry();
  private middlewareRegistry = new MiddlewareRegistry();

  private overridePolicy: OverridePolicy = OverridePolicy.FORBID;
  private currentPluginId: string = '';

  // ============================================================
  // Context Management
  // ============================================================

  /**
   * Set the plugin context for the current registration session.
   * Called by PluginEngine before invoking plugin.registerContributions().
   */
  setPluginContext(pluginId: string, overridePolicy?: OverridePolicy): void {
    this.currentPluginId = pluginId;
    if (overridePolicy !== undefined) {
      this.overridePolicy = overridePolicy;
    }
  }

  /**
   * Set the override policy for contribution conflicts.
   */
  setOverridePolicy(policy: OverridePolicy): void {
    this.overridePolicy = policy;
  }

  /**
   * Check if a contribution type is supported by this manager.
   */
  supportsType(type: ContributionType): boolean {
    switch (type) {
      case 'node-type':
      case 'tool-type':
      case 'llm-provider':
      case 'formatter':
      case 'event-handler':
      case 'hook-handler':
      case 'middleware':
        return true;
      default:
        return false;
    }
  }

  // ============================================================
  // ContributionRegistrar Implementation
  // ============================================================

  get nodeTypes(): NodeTypeRegistrar | undefined {
    return this.supportsType('node-type') ? this : undefined;
  }

  get toolTypes(): ToolTypeRegistrar | undefined {
    return this.supportsType('tool-type') ? this : undefined;
  }

  get llmProviders(): LLMProviderRegistrar | undefined {
    return this.supportsType('llm-provider') ? this : undefined;
  }

  get formatters(): FormatterRegistrar | undefined {
    return this.supportsType('formatter') ? this : undefined;
  }

  get eventHandlers(): EventHandlerRegistrar | undefined {
    return this.supportsType('event-handler') ? this : undefined;
  }

  get hookHandlers(): HookHandlerRegistrar | undefined {
    return this.supportsType('hook-handler') ? this : undefined;
  }

  get middleware(): MiddlewareRegistrar | undefined {
    return this.supportsType('middleware') ? this : undefined;
  }

  // ============================================================
  // NodeTypeRegistrar Implementation
  // ============================================================

  registerNodeType(type: string, handler: PluginNodeHandler): void {
    const nodeHandlerFn: NodeHandlerFn = async (_globalContext, _workflowExecutionEntity, _node, _context) => {
      const context = _context as Record<string, unknown> | undefined;
      const node = _node as unknown as Record<string, unknown> | undefined;
      const result = await handler.execute({
        nodeId: (node?.['id'] as string) ?? '',
        inputs: (context?.['inputs'] as Record<string, unknown>) ?? {},
        config: (node?.['config'] as Record<string, unknown>) ?? {},
      });
      return result.outputs;
    };
    this.registerNodeTypeInternal(this.currentPluginId, type, nodeHandlerFn);
  }

  // ============================================================
  // ToolTypeRegistrar Implementation
  // ============================================================

  registerToolType(type: string, executor: PluginToolExecutor): void {
    const executorConstructor = class implements IToolExecutor {
      async execute(
        _tool: Tool,
        parameters: Record<string, unknown>,
        _options?: ToolExecutionOptions,
        _executionId?: string,
        _context?: Record<string, unknown>,
      ): Promise<ToolExecutionResult> {
        const result = await executor.execute({ args: parameters ?? {} });
        return { result: result.result } as ToolExecutionResult;
      }
      getExecutorType(): string {
        return type;
      }
    };
    this.registerToolTypeInternal(this.currentPluginId, type, executorConstructor as unknown as IToolExecutorConstructor);
  }

  // ============================================================
  // LLMProviderRegistrar Implementation
  // ============================================================

  registerLLMProvider(provider: string, formatter: PluginLLMFormatter): void {
    this.registerLLMProviderInternal(this.currentPluginId, provider, toBaseFormatter(formatter));
  }

  // ============================================================
  // FormatterRegistrar Implementation
  // ============================================================

  registerFormatter(name: string, formatter: PluginLLMFormatter): void {
    this.registerFormatterInternal(this.currentPluginId, name, toBaseFormatter(formatter));
  }

  // ============================================================
  // EventHandlerRegistrar Implementation
  // ============================================================

  registerEventHandler(eventType: string, handler: PluginEventHandler): void {
    const eventHandler: EventHandler = (event: Record<string, unknown>) => {
      return handler.handle({ type: eventType, data: event });
    };
    this.registerEventHandlerInternal(this.currentPluginId, eventType, eventHandler);
  }

  // ============================================================
  // HookHandlerRegistrar Implementation
  // ============================================================

  registerHookHandler(hookType: string, handler: PluginHookHandler): void {
    const hookHandler: HookHandler = (context: Record<string, unknown>) => {
      return handler.handle(context);
    };
    this.registerHookHandlerInternal(this.currentPluginId, hookType, hookHandler);
  }

  // ============================================================
  // MiddlewareRegistrar Implementation
  // ============================================================

  registerMiddleware(_phase: string, middleware: PluginExecutionMiddleware): void {
    const phaseKey = middleware.phase as MiddlewarePhase;
    const executionMw: ExecutionMiddleware = {
      phase: phaseKey,
      handler: middleware.handler,
      priority: middleware.priority,
    };
    this.registerMiddlewareInternal(this.currentPluginId, phaseKey, executionMw);
  }

  // ============================================================
  // Internal Registration Methods
  // ============================================================

  private registerNodeTypeInternal(
    pluginId: string,
    nodeType: string,
    handler: NodeHandlerFn,
    template?: Record<string, unknown>,
  ): void {
    this.handleConflict('node-type', nodeType, pluginId, () => {
      this.nodeTypeRegistry.register(pluginId, nodeType, handler, template);
    });
  }

  private registerToolTypeInternal(pluginId: string, type: string, executor: IToolExecutorConstructor): void {
    this.handleConflict('tool-type', type, pluginId, () => {
      this.toolTypeRegistry.register(pluginId, type, executor);
    });
  }

  private registerLLMProviderInternal(pluginId: string, provider: string, formatter: BaseFormatter): void {
    this.handleConflict('llm-provider', provider, pluginId, () => {
      this.llmProviderRegistry.register(pluginId, provider, formatter);
    });
  }

  private registerFormatterInternal(pluginId: string, name: string, formatter: BaseFormatter): void {
    this.handleConflict('formatter', name, pluginId, () => {
      this.formatterRegistry.register(pluginId, name, formatter);
    });
  }

  private registerEventHandlerInternal(pluginId: string, eventType: string, handler: EventHandler, priority?: number): void {
    this.eventHandlerRegistry.register(pluginId, eventType, handler, priority);
  }

  private registerHookHandlerInternal(pluginId: string, hookType: string, handler: HookHandler): void {
    this.hookHandlerRegistry.register(pluginId, hookType, handler);
  }

  private registerMiddlewareInternal(_pluginId: string, phase: MiddlewarePhase, mw: ExecutionMiddleware): void {
    this.middlewareRegistry.register(phase, mw);
  }

  // ============================================================
  // Unregistration
  // ============================================================

  /**
   * Unregister all contributions from a plugin.
   */
  unregisterAll(pluginId: string): void {
    this.nodeTypeRegistry.unregisterByPluginId(pluginId);
    this.toolTypeRegistry.unregisterByPluginId(pluginId);
    this.llmProviderRegistry.unregisterByPluginId(pluginId);
    this.formatterRegistry.unregisterByPluginId(pluginId);
    this.eventHandlerRegistry.unregisterByPluginId(pluginId);
    this.hookHandlerRegistry.unregisterByPluginId(pluginId);
    this.middlewareRegistry.unregisterByPluginId(pluginId);

    this.logger.info(`Unregistered all contributions for plugin: ${pluginId}`);
  }

  // ============================================================
  // Query Methods
  // ============================================================

  getNodeHandler(nodeType: string): NodeHandlerFn | undefined {
    return this.nodeTypeRegistry.getHandler(nodeType);
  }

  getToolExecutor(type: string): IToolExecutorConstructor | undefined {
    return this.toolTypeRegistry.getExecutor(type);
  }

  getAllToolExecutors(): Map<string, { executor: IToolExecutorConstructor; pluginId: string }> {
    return this.toolTypeRegistry.getAllExecutors() as Map<string, { executor: IToolExecutorConstructor; pluginId: string }>;
  }

  getLLMProvider(provider: string): BaseFormatter | undefined {
    return this.llmProviderRegistry.getFormatter(provider);
  }

  getFormatter(name: string): BaseFormatter | undefined {
    return this.formatterRegistry.getFormatter(name);
  }

  hasNodeHandler(nodeType: string): boolean {
    return this.nodeTypeRegistry.hasHandler(nodeType);
  }

  getAllNodeTypes(): string[] {
    return this.nodeTypeRegistry.getAllNodeTypes();
  }

  getAllLLMProviders(): string[] {
    return this.llmProviderRegistry.getAllProviders();
  }

  getAllToolTypes(): string[] {
    return this.toolTypeRegistry.getAllToolTypes();
  }

  getAllEventHandlers(): Map<string, EventHandler> {
    return this.eventHandlerRegistry.getAllHandlers();
  }

  getAllHookHandlers(): Map<string, HookHandler> {
    return this.hookHandlerRegistry.getAllHandlers();
  }

  getMiddleware(phase: MiddlewarePhase): ExecutionMiddleware[] {
    return this.middlewareRegistry.getMiddleware(phase);
  }

  hasMiddleware(phase: MiddlewarePhase): boolean {
    return this.middlewareRegistry.hasMiddleware(phase);
  }

  /**
   * Execute all middleware for a given phase in sequence (onion model).
   * Silently skips if no middleware is registered for the phase.
   */
  async runMiddleware(phase: MiddlewarePhase, context: Record<string, unknown>): Promise<void> {
    await this.middlewareRegistry.runMiddleware(phase, context);
  }

  // ============================================================
  // Conflict Handling
  // ============================================================

  private handleConflict(
    type: ContributionType,
    key: string,
    pluginId: string,
    registerFn: () => void,
  ): void {
    const existing = this.getExistingEntry(type, key);
    if (existing && existing !== pluginId) {
      switch (this.overridePolicy) {
        case OverridePolicy.FORBID:
          this.logger.warn(
            `Plugin '${pluginId}' cannot override ${type} '${key}' (owned by '${existing}'). Override policy is FORBID.`,
          );
          return;
        case OverridePolicy.WARN:
          this.logger.warn(
            `Plugin '${pluginId}' is overriding ${type} '${key}' (was owned by '${existing}').`,
          );
          registerFn();
          return;
        case OverridePolicy.ALLOW:
        case OverridePolicy.PRIORITY:
          registerFn();
          return;
      }
    }
    registerFn();
  }

  private getExistingEntry(type: ContributionType, key: string): string | undefined {
    switch (type) {
      case 'node-type':
        return this.nodeTypeRegistry.getOwner(key);
      case 'tool-type':
        return this.toolTypeRegistry.getOwner(key);
      case 'llm-provider':
        return this.llmProviderRegistry.getOwner(key);
      case 'formatter':
        return this.formatterRegistry.getOwner(key);
      default:
        return undefined;
    }
  }
}

File: crates/wf-plugin/src/contributions/manager.rs
use std::sync::Arc;
use std::sync::RwLock;

use serde_json::Value;
use wf_types::{HookType, MiddlewarePhase};

use super::registrar::ContributionRegistrar;
use super::registries::{MultiRegistry, Registry};
use super::types::*;
use crate::error::{PluginError, PluginResult};

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub enum OverridePolicy {
    #[default]
    Forbid,
    Warn,
    Allow,
    Priority,
}

pub struct ContributionManager {
    current_plugin_id: RwLock<String>,
    override_policy: RwLock<OverridePolicy>,
    node_type_registry: Registry<String, Arc<dyn PluginNodeHandler>>,
    tool_type_registry: Registry<String, Arc<dyn PluginToolExecutor>>,
    llm_provider_registry: Registry<String, Arc<dyn PluginLLMFormatter>>,
    formatter_registry: Registry<String, Arc<dyn PluginLLMFormatter>>,
    event_handler_registry: MultiRegistry<String, Arc<dyn PluginEventHandler>>,
    hook_handler_registry: MultiRegistry<String, Arc<dyn PluginHookHandler>>,
    middleware_registry: MultiRegistry<String, (i32, Arc<dyn PluginMiddlewareHandler>)>,
}

impl Default for ContributionManager {
    fn default() -> Self {
        Self::new()
    }
}

impl ContributionManager {
    pub fn new() -> Self {
        Self {
            current_plugin_id: RwLock::new(String::new()),
            override_policy: RwLock::new(OverridePolicy::Forbid),
            node_type_registry: Registry::new(),
            tool_type_registry: Registry::new(),
            llm_provider_registry: Registry::new(),
            formatter_registry: Registry::new(),
            event_handler_registry: MultiRegistry::new(),
            hook_handler_registry: MultiRegistry::new(),
            middleware_registry: MultiRegistry::new(),
        }
    }

    pub fn set_override_policy(&self, policy: OverridePolicy) {
        *self.override_policy.write().unwrap() = policy;
    }

    pub fn start_registration(&self, plugin_id: &str) {
        *self.current_plugin_id.write().unwrap() = plugin_id.to_owned();
    }

    pub fn as_registrar(&self) -> RegistrarGuard<'_> {
        RegistrarGuard { manager: self }
    }

    pub fn unregister_all(&self, plugin_id: &str) {
        self.node_type_registry.unregister_by_plugin(plugin_id);
        self.tool_type_registry.unregister_by_plugin(plugin_id);
        self.llm_provider_registry.unregister_by_plugin(plugin_id);
        self.formatter_registry.unregister_by_plugin(plugin_id);
        self.event_handler_registry.unregister_by_plugin(plugin_id);
        self.hook_handler_registry.unregister_by_plugin(plugin_id);
        self.middleware_registry.unregister_by_plugin(plugin_id);
    }

    // --- Query methods ---

    pub fn get_node_handler(&self, type_name: &str) -> Option<Arc<dyn PluginNodeHandler>> {
        self.node_type_registry.get(type_name)
    }

    pub fn get_tool_executor(&self, type_name: &str) -> Option<Arc<dyn PluginToolExecutor>> {
        self.tool_type_registry.get(type_name)
    }

    pub fn get_llm_formatter(&self, name: &str) -> Option<Arc<dyn PluginLLMFormatter>> {
        self.llm_provider_registry.get(name)
    }

    pub fn get_formatter(&self, name: &str) -> Option<Arc<dyn PluginLLMFormatter>> {
        self.formatter_registry.get(name)
    }

    pub fn get_event_handlers(&self, event_type: &str) -> Vec<Arc<dyn PluginEventHandler>> {
        self.event_handler_registry.get(event_type)
    }

    pub fn get_hook_handlers(&self, hook_type: &HookType) -> Vec<Arc<dyn PluginHookHandler>> {
        self.hook_handler_registry.get(hook_type.as_str())
    }

    pub fn get_middleware(
        &self,
        phase: &MiddlewarePhase,
    ) -> Vec<(i32, Arc<dyn PluginMiddlewareHandler>)> {
        let mut handlers = self.middleware_registry.get(phase.as_str());
        handlers.sort_by_key(|(p, _)| *p);
        handlers
    }

    // --- Enumeration methods ---

    pub fn all_node_types(&self) -> Vec<(String, String)> {
        self.node_type_registry.all()
    }

    pub fn all_tool_types(&self) -> Vec<(String, String)> {
        self.tool_type_registry.all()
    }

    pub fn all_llm_providers(&self) -> Vec<(String, String)> {
        self.llm_provider_registry.all()
    }

    pub fn all_formatters(&self) -> Vec<(String, String)> {
        self.formatter_registry.all()
    }

    pub fn all_event_handlers(&self) -> Vec<(String, String)> {
        self.event_handler_registry.all()
    }

    pub fn all_hook_handlers(&self) -> Vec<(String, String)> {
        self.hook_handler_registry.all()
    }

    pub fn all_middleware_phases(&self) -> Vec<String> {
        self.middleware_registry.keys()
    }

    pub async fn run_middleware(&self, phase: &MiddlewarePhase, context: Value) -> PluginResult<()> {
        let handlers = self.get_middleware(phase);

        let mut next: NextFn = Box::new(|| Box::pin(async { Ok(()) }));

        for (_, handler) in handlers.into_iter().rev() {
            let prev = next;
            let ctx = context.clone();
            next = Box::new(move || {
                let h = handler;
                let p = prev;
                Box::pin(async move { h.handle(ctx, p).await })
            });
        }

        next().await
    }

    fn check_conflict(
        &self,
        type_name: &str,
        key: &str,
        owner_check: impl Fn() -> Option<String>,
    ) -> PluginResult<()> {
        let policy = *self.override_policy.read().unwrap();
        if let Some(owner) = owner_check() {
            let current = self.current_plugin_id.read().unwrap().clone();
            if owner != current {
                match policy {
                    OverridePolicy::Forbid => {
                        return Err(PluginError::ContributionConflict(format!(
                            "plugin '{}' cannot override {} '{}' (owned by '{}')",
                            current, type_name, key, owner
                        )));
                    }
                    OverridePolicy::Warn => {
                        tracing::warn!(
                            "plugin '{}' overriding {} '{}' (was '{}')",
                            current,
                            type_name,
                            key,
                            owner
                        );
                    }
                    OverridePolicy::Allow | OverridePolicy::Priority => {}
                }
            }
        }
        Ok(())
    }
}

pub struct RegistrarGuard<'a> {
    manager: &'a ContributionManager,
}

impl ContributionRegistrar for RegistrarGuard<'_> {
    fn register_node_type(&mut self, type_name: &str, handler: Arc<dyn PluginNodeHandler>) {
        let plugin_id = self.manager.current_plugin_id.read().unwrap().clone();
        if self
            .manager
            .check_conflict("node-type", type_name, || {
                self.manager.node_type_registry.get_owner(type_name)
            })
            .is_ok()
        {
            self.manager
                .node_type_registry
                .register(type_name.into(), plugin_id, handler);
        }
    }

    fn register_tool_type(&mut self, type_name: &str, executor: Arc<dyn PluginToolExecutor>) {
        let plugin_id = self.manager.current_plugin_id.read().unwrap().clone();
        if self
            .manager
            .check_conflict("tool-type", type_name, || {
                self.manager.tool_type_registry.get_owner(type_name)
            })
            .is_ok()
        {
            self.manager
                .tool_type_registry
                .register(type_name.into(), plugin_id, executor);
        }
    }

    fn register_llm_provider(&mut self, name: &str, formatter: Arc<dyn PluginLLMFormatter>) {
        let plugin_id = self.manager.current_plugin_id.read().unwrap().clone();
        if self
            .manager
            .check_conflict("llm-provider", name, || {
                self.manager.llm_provider_registry.get_owner(name)
            })
            .is_ok()
        {
            self.manager
                .llm_provider_registry
                .register(name.into(), plugin_id, formatter);
        }
    }

    fn register_formatter(&mut self, name: &str, formatter: Arc<dyn PluginLLMFormatter>) {
        let plugin_id = self.manager.current_plugin_id.read().unwrap().clone();
        if self
            .manager
            .check_conflict("formatter", name, || {
                self.manager.formatter_registry.get_owner(name)
            })
            .is_ok()
        {
            self.manager
                .formatter_registry
                .register(name.into(), plugin_id, formatter);
        }
    }

    fn register_event_handler(&mut self, event_type: &str, handler: Arc<dyn PluginEventHandler>) {
        let plugin_id = self.manager.current_plugin_id.read().unwrap().clone();
        self.manager
            .event_handler_registry
            .register(event_type.into(), plugin_id, handler);
    }

    fn register_hook_handler(&mut self, hook_type: HookType, handler: Arc<dyn PluginHookHandler>) {
        let plugin_id = self.manager.current_plugin_id.read().unwrap().clone();
        let key = hook_type.as_str().to_string();
        self.manager
            .hook_handler_registry
            .register(key, plugin_id, handler);
    }

    fn register_middleware(
        &mut self,
        phase: MiddlewarePhase,
        priority: i32,
        handler: Arc<dyn PluginMiddlewareHandler>,
    ) {
        let plugin_id = self.manager.current_plugin_id.read().unwrap().clone();
        let key = phase.as_str().to_string();
        self.manager
            .middleware_registry
            .register(key, plugin_id, (priority, handler));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::contributions::types::NextFn;

    struct NoopHookHandler;

    #[async_trait::async_trait]
    impl PluginHookHandler for NoopHookHandler {
        async fn handle(&self, _context: Value) -> PluginResult<()> {
            Ok(())
        }
    }

    struct NoopMiddlewareHandler;

    #[async_trait::async_trait]
    impl PluginMiddlewareHandler for NoopMiddlewareHandler {
        async fn handle(&self, _context: Value, _next: NextFn) -> PluginResult<()> {
            Ok(())
        }
    }

    #[test]
    fn arbitrary_hook_string_queryable_via_other() {
        let manager = ContributionManager::new();
        manager.start_registration("p1");
        let mut registrar = manager.as_registrar();
        registrar.register_hook_handler(
            HookType::Other("custom-hook".into()),
            Arc::new(NoopHookHandler),
        );
        assert_eq!(
            manager
                .get_hook_handlers(&HookType::Other("custom-hook".into()))
                .len(),
            1
        );
        // A registration under an arbitrary string does not match known types.
        assert!(manager.get_hook_handlers(&HookType::OnError).is_empty());
    }

    #[test]
    fn known_hook_string_round_trips_through_registry() {
        let manager = ContributionManager::new();
        manager.start_registration("p1");
        let mut registrar = manager.as_registrar();
        registrar.register_hook_handler(HookType::OnError, Arc::new(NoopHookHandler));
        assert_eq!(manager.get_hook_handlers(&HookType::OnError).len(), 1);
        assert_eq!(
            manager.get_hook_handlers(&HookType::from("on_error")).len(),
            1
        );
    }

    #[test]
    fn middleware_phase_round_trips_through_registry() {
        let manager = ContributionManager::new();
        manager.start_registration("p1");
        let mut registrar = manager.as_registrar();
        registrar.register_middleware(
            MiddlewarePhase::OnCheckpoint,
            10,
            Arc::new(NoopMiddlewareHandler),
        );
        assert_eq!(
            manager.get_middleware(&MiddlewarePhase::OnCheckpoint).len(),
            1
        );
        assert_eq!(
            manager
                .get_middleware(&MiddlewarePhase::from("on-checkpoint"))
                .len(),
            1
        );
        assert!(
            manager
                .get_middleware(&MiddlewarePhase::Other("custom-phase".into()))
                .is_empty()
        );
    }
}
